src/Features/Core/Portable/MakeFieldReadonly/MakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,254 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.MakeFieldReadonly
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal sealed class MakeFieldReadonlyDiagnosticAnalyzer
+        : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        public MakeFieldReadonlyDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.MakeFieldReadonlyDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Add_readonly_modifier), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Make_field_readonly), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                // State map for fields:
+                //  'isCandidate' : Indicates whether the field is a candidate to be made readonly based on it's declaration and options.
+                //  'written'     : Indicates if there are any writes to the field outside the constructor and field initializer.
+                var fieldStateMap = new ConcurrentDictionary<IFieldSymbol, (bool isCandidate, bool written)>();
+
+                // We register following actions in the compilation:
+                // 1. A symbol action for field symbols to ensure the field state is initialized for every field in the compilation.
+                // 2. An operation action for field references to detect if a candidate field is written outside constructor and field initializer, and update field state accordingly.
+                // 3. A symbol start/end action for named types to report diagnostics for candidate fields that were not written outside constructor and field initializer.
```

Comment:
I personally would appreciate wrapping long comments that don't show up well in github.  Usually around 120 columns is a good point.  It means things show up fine on 1080p screens, while still giving enough space for the comment to not feel cramped.